File: src/models/person.py
import numpy as np
import random
from engine.circle import Circle
from engine.drawable import Drawable
from engine.utils.vector_ops import *
from engine.utils.constants import *

COLOR_MAP = {
    "S": BLUE,
    "I": RED,
    "R": GREY
}

class Person:

    def __init__(self, x=0, y=0, r = 5.0):
        # Visuals
        self.color = WHITE
        self.visual = Circle(0, 0, r)
        self.visual.color = self.color
        self.gravity_well_visual = Circle(0, 0, 3)

        # Time
        self.time = 0
        self.last_step_change = -1

        # Bounds
        self.dl_bound = np.array((0, 0))
        self.ur_bound = np.array((540, 480))

        # Movement
        self.position = np.array((x, y, 0))
        self.velocity = np.zeros(3)
        self.max_speed = 30
        self.gravity_well = None
        self.gravity_strength = 500000
        self.wander_step_size = 100
        self.wander_step_duration = 2
        self.wall_buffer = 20
        self.wall_strength = 10000

        # Infection
        self.status = "S" # SRI status
        self.infection_radius = 15
        self.infection_start_time = np.inf
        self.infection_end_time = np.inf
        self.num_infected = 0
        self.city = None
    
    def update(self, dt):
        total_force = np.zeros(3)
        self.time += dt
        # Wander
        if self.wander_step_size != 0:
            if (self.time - self.last_step_change) > self.wander_step_duration:
                vect = z_rotation(RIGHT, TAU * random.random())
                self.gravity_well = self.position + (self.wander_step_size * vect)
                self.last_step_change = self.time
                # DEBUG: Gravity well visual
                self.gravity_well_visual.position = self.gravity_well

        if self.gravity_well is not None:
            to_well = (self.gravity_well - self.position)
            dist = get_norm(to_well)
            if dist != 0:
                total_force += self.gravity_strength * to_well / (dist**3)
        
        # TODO: Social distance

        # TODO: Avoid walls
        wall_force = np.zeros(3)
        for i in range(2):
            to_lower = self.position[i] - self.dl_bound[i]
            to_upper = self.ur_bound[i] - self.position[i]

            # Bounce
            if to_lower < 0:
                self.velocity[i] = np.abs(self.velocity[i])
                self.position[i] = self.dl_bound[i]
            if to_upper < 0:
                self.velocity[i] = -np.abs(self.velocity[i])
                self.position[i] = self.ur_bound[i]
            
            wall_force[i] += max((-1 / self.wall_buffer + 1 / to_lower), 0)
            wall_force[i] -= max((-1 / self.wall_buffer + 1 / to_upper), 0)
        total_force += wall_force * self.wall_strength

        # Apply force
        self.velocity += total_force * dt

        # Limit speed
        speed = get_norm(self.velocity)
        if speed > self.max_speed:
            self.velocity *= self.max_speed / speed
        
        self.shift(self.velocity * dt)

    def set_bounds(self, dl, ur):
        self.dl_bound = dl
        self.ur_bound = ur

    def shift(self, vector):
        self.position += vector
        self.visual.position = self.position
        return self
    
    def set_color(self, color):
        self.visual.color = color
    
    def render(self):
        self.visual.render()
        # DEBUG: Draw gravity well
        #self.gravity_well_visual.render()

File: src/MonoGame/SirSimulation/Models/Person.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using SirSimulation.Engine;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MonoGame.Extended;

namespace SirSimulation.Models
{
    public class Person : Engine.IDrawable
    {
        private readonly Texture2D _sSprite;
        private readonly Texture2D _iSprite;
        private readonly Texture2D _rSprite;

        private Texture2D _currentSprite;
        private double _lastStepChange = -1;

        public Person(Texture2D sSprite, Texture2D iSprite, Texture2D rSprite)
        {
            _sSprite = sSprite;
            _iSprite = iSprite;
            _rSprite = rSprite;

            _currentSprite = _sSprite;
            Status = InfectionStatus.Susceptible;
        }

        public Vector2 Position { get; set; } = Vector2.Zero;
        public Vector2 Velocity { get; set; } = Vector2.Zero;
        public float MaxSpeed { get; set; } = 40;
        public Rectangle Bounds { get; set; }
        public Vector2 GravityWell { get; set; } = Vector2.Zero;
        public float GravityStrength { get; set; } = 600000;
        public float WanderStepSize { get; set; } = 140;
        public double WanderStepDuration { get; set; } = 2;
        public float WallBuffer { get; set; } = 20;
        public float WallStrength { get; set; } = 10000;

        public InfectionStatus Status { get; private set; }
        public float InfectionRadius { get; set; }
        public double InfectionStartTime { get; private set; } = float.PositiveInfinity;
        public double InfectionEndTime { get; private set; } = float.PositiveInfinity;
        public int NumInfected { get; set; } = 0;
        public City City { get; set; }

        public void SetStatus(InfectionStatus newStatus, GameTime gameTime)
        {
            switch (newStatus)
            {
                case InfectionStatus.Susceptible:
                    _currentSprite = _sSprite;
                    break;
                case InfectionStatus.Infected:
                    _currentSprite = _iSprite;
                    InfectionStartTime = gameTime.TotalGameTime.TotalSeconds;
                    break;
                case InfectionStatus.Removed:
                    _currentSprite = _rSprite;
                    InfectionEndTime = gameTime.TotalGameTime.TotalSeconds;
                    break;
                default:
                    break;
            }
            Status = newStatus;
        }

        public void Update(TimeSpan totalTime, TimeSpan elapsedTime)
        {
            var totalForce = Vector2.Zero;

            // Wander
            if (WanderStepSize != 0)
            {
                if (totalTime.TotalSeconds - _lastStepChange > WanderStepDuration)
                {
                    var vect = Constants.Right.Rotate(MathHelper.TwoPi * Utils.Random.NextFloat());
                    GravityWell = Position + (vect * WanderStepSize);
                    _lastStepChange = totalTime.TotalSeconds;
                }
            }

            if (GravityWell != null)
            {
                var toWell = GravityWell - Position;
                var dist = Vector2.Distance(GravityWell, Position);
                if (dist > 0)
                {
                    totalForce += GravityStrength * toWell / ((float)Math.Pow(dist, 3));
                }
            }

            // Avoid walls
            float wallForceX = 0;
            float wallForceY = 0;
            var toRight = Bounds.Right - Position.X;
            if(toRight < 0)
            {
                Velocity = new Vector2(Math.Abs(Velocity.X), Velocity.Y);
            }
            wallForceX -= Math.Max(-1f / WallBuffer + 1f / toRight, 0);
            var toLeft = Position.X - Bounds.Left;
            if(toLeft < 0)
            {
                Velocity = new Vector2(Math.Abs(Velocity.X), Velocity.Y);
            }
            wallForceX += Math.Max(-1f / WallBuffer + 1f / toLeft, 0);
            var toTop = Position.Y - Bounds.Top;
            if(toTop < 0)
            {
                Velocity = new Vector2(Velocity.X, Math.Abs(Velocity.Y));
            }
            wallForceY += Math.Max(-1f / WallBuffer + 1f / toTop, 0);
            var toBottom = Bounds.Bottom - Position.Y;
            if(toBottom < 0)
            {
                Velocity = new Vector2(Velocity.X, -Math.Abs(Velocity.Y));
            }
            wallForceY -= Math.Max(-1f / WallBuffer + 1f / toBottom, 0);
            var wallForce = new Vector2(wallForceX, wallForceY);
            totalForce += wallForce * WallStrength;

            // Limit speed
            var speed = Velocity.Normal();
            if (speed > MaxSpeed)
            {
                Velocity *= MaxSpeed / speed;
            }

            var dt = (float)elapsedTime.TotalSeconds;

            Velocity += totalForce * dt;

            Position += Velocity * dt;
        }

        public void Draw(GameTime gameTime, SpriteBatch spriteBatch)
        {
            var pos = Position - new Vector2(_sSprite.Width / 2, _sSprite.Height / 2);
            spriteBatch.Draw(_currentSprite, pos, Color.White);
        }

    }
}
